File: joaat.c
#include <stdint.h>

uint32_t joaat(const char * s){
	uint32_t hash = 0;

	for (; *s; ++s){
		hash += *s;
		hash += (hash << 10);
		hash ^= (hash >> 6);
	}

	hash += (hash << 3);
	hash ^= (hash >> 11);
	hash += (hash << 15);

	return hash;
}

File: joaat.php
<?php

function int32($i){
	return $i & 0xFFFFFFFF;
}

function signedInt32($i){
	return $i | (-($i & 0x80000000));
}

function signedInt8($i){
	return $i | (-($i & 0x80));
}

function joaat($str){
	$hash = 0;
	foreach(str_split(mb_strtolower($str)) as $c){
		$hash = int32($hash + signedInt8(ord($c)));
		$hash = int32($hash + ($hash << 10));
		$hash = int32($hash ^ ($hash >> 6));
	}
	$hash = int32($hash + ($hash << 3));
	$hash = int32($hash ^ ($hash >> 11));
	$hash = int32($hash + ($hash << 15));
	
	return [
		"hex" => strtoupper(dechex($hash)),
		"unsigned" => $hash,
		"signed" => signedInt32($hash),
	];
}

?>